feature/subscriber-attributes/src/main/java/com/revenuecat/purchases/subscriberattributes/caching/SubscriberAttributesCache.kt

```diff
@@ -0,0 +1,120 @@
+
+import com.revenuecat.purchases.common.caching.DeviceCache
+import com.revenuecat.purchases.common.debugLog
+import com.revenuecat.purchases.subscriberattributes.SubscriberAttribute
+import com.revenuecat.purchases.subscriberattributes.buildSubscriberAttributesMapPerUser
+
+internal typealias AppUserID = String
+internal typealias SubscriberAttributeMap = Map<String, SubscriberAttribute>
+internal typealias SubscriberAttributesPerAppUserIDMap = Map<AppUserID, SubscriberAttributeMap>
+
+class SubscriberAttributesCache(
+    internal val deviceCache: DeviceCache
+) {
+    internal val subscriberAttributesCacheKey: String by lazy {
+        deviceCache.newKey("subscriberAttributes")
+    }
+
+    // region subscriber attributes
+
+    @Synchronized
+    fun setAttributes(appUserID: AppUserID, attributesToBeSet: SubscriberAttributeMap) {
+        val currentlyStoredAttributesForAllUsers = getAllStoredSubscriberAttributes()
+        val currentlyStoredAttributesForAppUserID =
+            currentlyStoredAttributesForAllUsers[appUserID] ?: emptyMap()
+        val updatedAttributesForUser = currentlyStoredAttributesForAppUserID + attributesToBeSet
+        val updatedAttributesForAllUsers =
+            currentlyStoredAttributesForAllUsers + mapOf(appUserID to updatedAttributesForUser)
+
+        deviceCache.putAttributes(updatedAttributesForAllUsers)
+    }
+
+    @Synchronized
+    fun getAllStoredSubscriberAttributes(): SubscriberAttributesPerAppUserIDMap =
+        deviceCache.getJSONObjectOrNull(subscriberAttributesCacheKey)
+            ?.buildSubscriberAttributesMapPerUser() ?: emptyMap()
+
+    @Synchronized
+    fun getAllStoredSubscriberAttributes(appUserID: AppUserID): SubscriberAttributeMap =
+        getAllStoredSubscriberAttributes()[appUserID] ?: emptyMap()
+
+    @Synchronized
+    fun getUnsyncedSubscriberAttributes(): SubscriberAttributesPerAppUserIDMap =
+        getAllStoredSubscriberAttributes()
+            .mapValues { (appUserID, attributesForUser) ->
+                attributesForUser.filterUnsynced(appUserID)
+            }.filterValues { it.isNotEmpty() }
+
+    @Synchronized
+    fun getUnsyncedSubscriberAttributes(appUserID: String) =
+        getAllStoredSubscriberAttributes(appUserID).filterUnsynced(appUserID)
+
+    private fun SubscriberAttributeMap.filterUnsynced(appUserID: AppUserID): SubscriberAttributeMap =
+        this.filterValues { !it.isSynced }
+            .also { unsyncedAttributesByKey ->
+                debugLog("Found ${unsyncedAttributesByKey.count()} unsynced attributes for appUserID: $appUserID \n" +
+                    if (unsyncedAttributesByKey.isNotEmpty()) {
+                        unsyncedAttributesByKey.values.joinToString("\n")
+                    } else {
+                        ""
+                    }
+                )
+            }
+
+    @Synchronized
+    fun clearSubscriberAttributesIfSyncedForSubscriber(appUserID: String) {
```

Comment:
maybe we could put the public stuff up top, internal in the middle and private at the bottom